src/NLog/Contexts/MappedDiagnosticsLogicalContext.cs

```diff
@@ -389,7 +361,9 @@ public static IDisposable SetScoped(IReadOnlyList<KeyValuePair<string,object>> i
                 IReadOnlyList<KeyValuePair<string, object>> itemsWithPreviousValues = null;
                 if (restorePreviousValueOnDispose)
                 {
-                    itemsWithPreviousValues = items.Select(x => new KeyValuePair<string, object>(x.Key, GetObject(x.Key, logicalContext))).ToArray();
+                    itemsWithPreviousValues = items
```

Comment:
Ah, right. First look tells me that it can be fixed with `.Where(x => logicalContext.Contains(x))`.
I'll check that and also with single key approach.